### DISCUSSION TITLE
Segmentation fault occurs in program running

### DISCUSSION THREAD

**Discussion by xchengood** [idaholab/moose#20318] [Q&A Modules: Solid mechanics] [answered]
Hi, Moose experts or users,

Recently, I take one creep model for reference ('A physics-based crystallographic modeling framework for describing the thermal creep behavior of Fe-Cr alloys') and try to write an individual class including .h and .c files. After debugging, there are no errors in compiling. When attempting to run the program with .i extension, I got the 'Segmentation fault' message below. 
[error.txt](https://github.com/idaholab/moose/files/8082026/error.txt)

So two questions here:
1. What possible reasons cause this? And have any suggestions to deal with it?
2. Are there good ways to check, debug or verify the codes?

Appreciate your help.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Since you are using a custom class, there are a plethora of reasons you could hit a seg fault. Some simple (reading out of bounds for example), some more complex (using data that has not been initialized for example)
The good way to work on this is to follow our debugging instructions, located here:
https://mooseframework.inl.gov/application_development/debugging.html

Guillaume

**Reply by xchengood** [reply]
Hi @GiudGiud. Thank you for your useful information. I followed debugging instructions to debug my codes and got the following errors. Could you have a look and what errors do those messages indicate? Thank you.
![Screen Shot 2022-02-21 at 5 10 17 PM](https://user-images.githubusercontent.com/74226455/155036573-45997d3a-3144-4dc5-97dd-2b0ff6f15913.png)

**Reply by GiudGiud** [reply]
I dont see anything obvious sadly. 
It is seg-faulting while trying to access an array, but I dont know why. 
In the debugger, can you print local variables see if anything is odd?

**Reply by loganharbour** [reply]
Looks like you're indexing into an array out of bounds in `CrystalPlasticityOrowanDislocationSlip.C` on line 519. We need to see this class to help more.

**Reply by xchengood** [reply]
Hi @GiudGiud and @loganharbour,
Thank you for your help. Now I fixed all issues and the .i file can be successfully run. I still have two questions.
1. The base class I used in the program is 'CrystalPlasticityStressUpdateBase'. I can get data on strain and stress. I am not sure if the obtained strain is the creep strain. Studying the curve of creep strain vs. time is my purpose.
2. Usually, in order to study the curve of creep strain vs. time well, should I keep running the program for at least several hours?
Appreciate your help again.

**Reply by GiudGiud** [reply]
@jiangwen84 will know about these tensor mechanics things

**Reply by sapitts** [reply]
The crystal plasticity 'stress calculator' and base class are described in detail on the documentation page: https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeMultipleCrystalPlasticityStress.html . 

1. The total Lagrangian strain is calculated in the stress calculator class (ComputeMultipleCrystalPlasticityStress), and you can use the plastic deformation gradient to determine when plastic deformation begins. 
2. Generally crystal plasticity simulations require a small timestep size and have many stateful properties. Both of these factor will contribute to a longer simulation run time.

**Reply by xchengood** [reply]
Hi @sapitts.

Thank you for your helpful information. 

1. Are there developed examples about the determination of plastic deformation beginning for reference?
2. I saw there is a class called 'RadialReturnCreepStressUpdateBase' inherited from 'RadialReturnStressUpdate' in which creep strain is determined. Class named 'PowerLawCreepStressUpdate' inherits this class well. Do you have any thoughts or suggestions about how to apply these classes to my case (the base class is CrystalPlasticityStressUpdateBase) making me easier and simpler to determine creep strain?

Thank you.

**Reply by sapitts** [reply]
1. I usually watch the plastic deformation gradient tensor diagonal components to determine the onset of plasticity: when any of the diagonal components change from unity (e.g. 0.9999999 or 1.000011), then I consider the strain from then on to be the inelastic strain.
2. I'd strongly recommend staying with the `CrystalPlasticityStressUpdateBase` as your base class. The other classes you've listed assume a J2 yield surface which is not appropriate for a crystal plasticity model.  In the crystal plasticity classes we don't really distinguish between plastic strain and creep strain but rather consider all of them inelastic strain and inelastic deformation

**Reply by xchengood** [reply]
Hi @sapitts. Thank you for your suggestions very much.
